From 2c7a1f1ffc03040c1f962e64d3741c5ed224e0f7 Mon Sep 17 00:00:00 2001
From: Konstantin <ria.freelander@gmail.com>
Date: Mon, 4 Jul 2022 17:02:29 +0300
Subject: [PATCH] xwayland: fix glamor es black windows

On some drivers (mesa included), when we request GBM_FORMAT_XRGB8888,
driver set surface storage format to GL_RGB8, which breaks gles2 rendering.
If we force set gbm_format to GBM_FORMAT_ARGB8888, then rendering
will happen and working. For correct colors, we need !924.

Fixes #1288
Fixes #1356

Signed-off-by: Konstantin <ria.freelander@gmail.com>
---
 hw/xwayland/xwayland-glamor-gbm.c | 11 +++++++----
 hw/xwayland/xwayland-screen.h     |  1 +
 2 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/hw/xwayland/xwayland-glamor-gbm.c b/hw/xwayland/xwayland-glamor-gbm.c
index 04cb5bc23d..412d024745 100644
--- a/hw/xwayland/xwayland-glamor-gbm.c
+++ b/hw/xwayland/xwayland-glamor-gbm.c
@@ -83,12 +83,14 @@ xwl_gbm_get(struct xwl_screen *xwl_screen)
 }
 
 static uint32_t
-gbm_format_for_depth(int depth)
+gbm_format_for_depth(int depth, int gles)
 {
     switch (depth) {
     case 16:
         return GBM_FORMAT_RGB565;
     case 24:
+        if (gles)
+            return GBM_FORMAT_ARGB8888;
         return GBM_FORMAT_XRGB8888;
     case 30:
         return GBM_FORMAT_ARGB2101010;
@@ -288,7 +290,7 @@ xwl_glamor_gbm_create_pixmap(ScreenPtr screen,
         (hint == CREATE_PIXMAP_USAGE_BACKING_PIXMAP ||
          hint == CREATE_PIXMAP_USAGE_SHARED ||
          (xwl_screen->rootless && hint == 0))) {
-        uint32_t format = gbm_format_for_depth(depth);
+        uint32_t format = gbm_format_for_depth(depth, xwl_screen->glamor_gles);
         Bool implicit = FALSE;
 
 #ifdef GBM_BO_WITH_MODIFIERS
@@ -648,7 +650,7 @@ glamor_pixmap_from_fds(ScreenPtr screen, CARD8 num_fds, const int *fds,
        data.width = width;
        data.height = height;
        data.num_fds = num_fds;
-       data.format = gbm_format_for_depth(depth);
+       data.format = gbm_format_for_depth(depth, xwl_screen->glamor_gles);
        data.modifier = modifier;
        for (i = 0; i < num_fds; i++) {
           data.fds[i] = fds[i];
@@ -665,7 +667,7 @@ glamor_pixmap_from_fds(ScreenPtr screen, CARD8 num_fds, const int *fds,
        data.width = width;
        data.height = height;
        data.stride = strides[0];
-       data.format = gbm_format_for_depth(depth);
+       data.format = gbm_format_for_depth(depth, xwl_screen->glamor_gles);
        bo = gbm_bo_import(xwl_gbm->gbm, GBM_BO_IMPORT_FD, &data,
                           GBM_BO_USE_RENDERING);
        implicit = TRUE;
@@ -1118,6 +1120,7 @@ xwl_glamor_gbm_init_egl(struct xwl_screen *xwl_screen)
     /* Mesa uses "drm" as backend name, in that case, just do nothing */
     if (gbm_backend_name && strcmp(gbm_backend_name, "drm") != 0)
         xwl_screen->glvnd_vendor = gbm_backend_name;
+    xwl_screen->glamor_gles = !epoxy_is_desktop_gl();
 
     return TRUE;
 error:
diff --git a/hw/xwayland/xwayland-screen.h b/hw/xwayland/xwayland-screen.h
index fadd0526e2..ff706641ee 100644
--- a/hw/xwayland/xwayland-screen.h
+++ b/hw/xwayland/xwayland-screen.h
@@ -61,6 +61,7 @@ struct xwl_screen {
 
     int rootless;
     int glamor;
+    int glamor_gles;
     int present;
     int force_xrandr_emulation;
     int fullscreen;
-- 
GitLab

